doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+# Materialize Formalism
+
+Materialize is a system that maintains views over changing data.
+
+At Materialize's heart are *time-varying collections* (TVCs).
+These are multisets of typed *data* whose contents may vary arbitrarily as function of some ordered *time*.
+Materialize records, transforms, and presents the contents of time varying collections.
+
+Materialize's primary function is to provide the contents of time-varying collection at specific times with absolute certainty.
+
+# Time-varying collections
+
+Materialize represents TVCs as a set of "updates" of the form `(data, time, diff)`.
+The difference `diff` represents the change in the multiplicity of `data` as we go from `time-1` to `time`.
+To go from a sequence of collections to this representation one subtracts the old from new frequencies for each `data`.
+To go from this representation to a sequence of collections, one updates the multiplicities of each `data` by `diff` for each `time`.
+
+Materialize most often only has partial information about TVCs.
+A TVC's future is usually not yet completely known, and its deep past also may not be faithfully recorded.
+To represent what Materialize knows with certainty, Materialize maintains two "frontiers" for each TVC.
+A frontier is a lower bound on times, and is used to describe the set of times greater or equal to the lower bound.
+The two frontiers Materialize maintains for each TVC are:
+
+*   The `since` frontier, which communicates that for times `time` greater or equal to `since`, the accumulated collection at `time` is precise.
+
+    Informally, updates at times not greater than the `since` frontier may have been consolidated with other updates.
```

Comment:
nit:

> may have been consolidated with other updates into a collection representing the state at the time represented by the frontier.

makes a stronger connection with the dichotomy between TVCs and collection sequences.

Reply:
I relaxed this a bit, because there is no guarantee that the advanced updates will be at `since`; they move in that direction, but e.g. may lag behind because we don't advance all compacted times when we advance `since`. Also, just trying to move the focus away from "what is the definition of times not greater or equal to `since`".